9mount.1
.TH "9mount" "1" "23 July 2008" "9mount" "User commands"
.SH NAME
9mount, 9bind, 9umount \- mount/unmount 9p filesystems
.SH SYNOPSIS
.B 9mount
[ insuvx ] [ -a SPEC ] [ -c CACHE ] [ -d DEBUG ] [ -m MSIZE ] DIAL MOUNTPT
.PP
.B 9bind
OLD NEW
.PP
.B 9umount
MOUNTPT
.SH DESCRIPTION
.B 9mount
mounts a 9p filesystem served at DIAL on MOUNTPT. MOUNTPT must be
writable by you and not sticky. DIAL is a dial string assuming one of
the forms:
.PP
unix!SOCKET
.br
tcp!HOST[!PORT]
.br
virtio!CHANNEL
.br
-
.PP
where SOCKET is the name of a file representing a socket, HOST is a
hostname, PORT is a port number or service name, and CHANNEL is a
virtio channel name (currently ignored). - indicates that 9p messages
should be read/written on stdin/stdout.
.B 9mount
has several options:
.TP
-i
mount the file system with your uid/gid
.TP
-n
dry-run, print mount command to stderr but don't actually mount anything
.TP
-s
single attach mode - all users accessing the mount point see the same
filesystem (by default they'll each see a unique attach)
.TP
-u
use the 9P2000.u extensions
.TP
-v
use device mapping
.TP
-x
exclusive access - other users cannot access the mount point
.TP
-a SPEC
SPEC determines which file tree to mount when attaching to file servers that
export multiple trees
.TP
-c CACHE
turns on caching using CACHE mode. Supported modes are
.I loose
(suitable for exclusive read-only mounts), 
.IR fscache ,
and
.IR mmap .
.TP
-d DEBUG
comma seperated list of channels for which to enable debug output. Possible
channels include: err, devel, 9p, vfs, conv, mux, trans, alloc, fcall.
.TP
-m MSIZE
specifies the maximum length of a single 9p message in bytes. Must be less
than or equal to 8192 for non-root users.
.PP
.B 9bind
performs a bind mount, making the tree visible at directory OLD also visible
at mount point NEW.
.PP
.B 9umount
unmounts a 9p filesystem previously mounted by you.
.SH ENVIRONMENT
.TP
$USER
the uname to provide to the server.
.SH EXAMPLES
.TP
9mount -i 'unix!/tmp/ns.'$USER'.:0/factotum' $HOME/n/factotum
mount p9p's factotum interface
.TP
9mount 'tcp!sources.cs.bell-labs.com' $HOME/n/sources
import plan 9's "sources"
.TP
9mount -u -a/home/sqweek/mail 'tcp!wren!5640' $HOME/mail
import my maildir from my server(wren), being served by ufs
.TP
9mount -i 'tcp!wren' $HOME/n/wren; 9bind $HOME/n/wren/home/sqweek/mail $HOME/mail
again importing my maildir, this time serving via u9fs
.SH BUGS
.B 9mount
truncates user names and SPECs to 249 characters.
.B 9umount
doesn't know this, so you won't be able to unmount anything
outside your home directory. But you probably never bother
logging out if your user name is that long.
.PP
.B 9mount
doesn't update /etc/mtab.
.PP
.B 9bind
only does a "shallow", non-recursive bind - any mounted filesystems
under the OLD tree will not appear mounted in the NEW tree.
.PP
If you
.B 9bind
a non-9p filesystem outside your home directory,
.B 9umount
won't let you unmount it.
.SH AUTHOR
sqweek@gmail.com
.SH SEE ALSO
.BR mount (1)


ChangeLog
2008-07-22 sqweek <sqweek@gmail.com>
	* 1.3 release
	* 9mount: support for custom uname, cache, msize, virtio, fd transport,
	access modes, -d more useful
	* 9umount: allow unmounting of even non-9p filesystems in your home dir
	* introduced some test cases
	* added ISC license

2008-06-20 sqweek <sqweek@gmail.com>
	* 1.2 release
	* 9mount: support for linux kernels >= 2.6.24
	* 9umount: allow unmounting of anything in your home dir

2007-10-16 sqweek <sqweek@gmail.com>
	* 1.1 release
	* 9bind: new SUID utility for bind mounting
	* Makefile: actually install the man page
	* 9mount: check that the initial calloc actually succeeds

2007-09-08 sqweek <sqweek@gmail.com>
	* 1.0 release


9mount.c
/* © 2008 sqweek <sqweek@gmail.com>
 * See COPYING for details.
 */
#include <err.h>
#include <mntent.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <limits.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/mount.h>
#include <arpa/inet.h>
#include <pwd.h>
#include <netdb.h>

#define nelem(x) (sizeof(x)/sizeof(*(x)))

enum {
	Maxmsize = 8192,
};

struct {char *mnemonic; int mask;} debug_flags[] = {
	{"err", 0x001},
	{"devel", 0x002},
	{"9p", 0x004},
	{"vfs", 0x008},
	{"conv", 0x010},
	{"mux", 0x020},
	{"trans", 0x040},
	{"alloc", 0x080},
	{"fcall", 0x100}
};

char*
append(char **dest, char *src, int *destlen)
{
	while (strlen(*dest) + 1 + strlen(src) > *destlen)
		*destlen *= 2;
	if (!(*dest=realloc(*dest, *destlen)))
		errx(1, "out of memory");

	if (**dest)
		strcat(*dest, ",");
	strcat(*dest, src);
	return *dest;
}

char*
getarg(char opt, char *cp, char*** argv)
{
	if (*(cp+1)) {
		return cp+1;
	} else if (*(*argv+1)) {
		return *++(*argv);
	} else {
		errx(1, "-%c: expected argument", opt);
	}
	return NULL;
}

void
parsedial(char *dial, char **network, char **netaddr, int *port)
{
	char *cp;
	if (!(*network=strtok(dial, "!"))) {
		errx(1, "empty dial string");
	}
	if (strcmp(*network, "unix") != 0
	&& strcmp(*network, "tcp") != 0
	&& strcmp(*network, "virtio") != 0) {
		errx(1, "%s: unknown network (expecting unix, tcp or virtio)", *network);
	}
	if (!(*netaddr=strtok(NULL, "!"))) {
		errx(1, "missing dial netaddress");
	}
	if (strcmp(*network, "tcp") == 0) {
		char *service;
		if ((service=strtok(NULL, "!"))) {
			if (strspn(service, "0123456789") == strlen(service)) {
				*port = atoi(service);
			} else {
				struct servent *sv;
				if ((sv=getservbyname(service, *network))) {
					/* sv->s_port is a 16-bit big endian masquerading as an int */
					*port = ntohs((uint16_t)sv->s_port);
					endservent();
				} else {
					errx(1, "%s: unknown service", service);
				}
			}
		}
	}
	if ((cp=strtok(NULL, "!"))) {
		errx(1, "%s: junk trailing dial string", cp);
	}
	if (strcmp(*network, "unix") == 0) {
		if (access(*netaddr, R_OK | W_OK)) {
			err(1, "%s", *netaddr);
		}
	}
}

int
main(int argc, char **argv)
{
	char buf[256], *opts, *dial = NULL, *mountpt = NULL;
	int optlen = 64, port = 0, i;
	struct stat stbuf;
	struct passwd *pw;
	int axess = 0, dotu = 0, uidgid = 0, dev = 0, debug = 0, dryrun = 0;
	char *debugstr = NULL, *msize = NULL, *cache = NULL, *aname = NULL;
	char *cp, *proto, *addr;

	if (!(opts=calloc(optlen, 1))) {
		err(1, "calloc");
	}
	while (*++argv) {
		if (**argv == '-' && (*argv)[1] != '\0') {
			for (cp=*argv+1; *cp; ++cp) {
				switch (*cp) {
					case 'i': uidgid = 1; break;
					case 'n': dryrun = 1; break;
					case 's': axess = -1; break;
					case 'u': dotu = 1; break;
					case 'v': dev = 1; break;
					case 'x': axess = getuid(); break;
					case 'a':
						aname = getarg('a', cp, &argv);
						*cp-- = '\0'; /* breaks out of for loop */
						break;
					case 'c':
						cache = getarg('c', cp, &argv);
						*cp-- = '\0';
						break;
					case 'd':
						debugstr = getarg('d', cp, &argv);
						*cp-- = '\0';
						break;
					case 'm':
						msize = getarg('m', cp, &argv);
						*cp-- = '\0';
						break;
					default:
						errx(1, "unrecognised argument '%c'", *cp);
				}
			}
		} else if (!dial) {
			dial = *argv;
		} else if (!mountpt) {
			mountpt = *argv;
		} else {
			errx(1, "%s: too many arguments", *argv);
		}
	}

	if (!dial || !mountpt) {
		errx(1, "usage: 9mount [ -insuvx ] [ -a spec ] [ -c cache ] [ -d debug ] [ -m msize ] dial mountpt");
	}

	if(!(pw=getpwuid(getuid()))) {
		err(1, "who are you?? getpwuid failed");
	}
	/* Make sure mount exists, is writable, and not sticky */
	if (stat(mountpt, &stbuf) || access(mountpt, W_OK)) {
		err(1, "%s", mountpt);
	}
	if (stbuf.st_mode & S_ISVTX) {
		errx(1, "%s: refusing to mount over sticky directory", mountpt);
	}

	if (strcmp(dial, "-") == 0) {
		proto = "fd";
		addr = "nodev";
		append(&opts, "rfdno=0,wfdno=1", &optlen);
	} else {
		parsedial(dial, &proto, &addr, &port);
	}

	/* set up mount options */
	append(&opts, proto, &optlen); /* < 2.6.24 */
	snprintf(buf, sizeof(buf), "trans=%s", proto);
	append(&opts, buf, &optlen); /* >= 2.6.24 */

	if (aname) {
		if (strchr(aname, ',')) {
			errx(1, "%s: spec can't contain commas", aname);
		}
		snprintf(buf, sizeof(buf), "aname=%s", aname);
		append(&opts, buf, &optlen);
	}

	if (cache) {
		if (strcmp(cache, "loose") != 0
		&& strcmp(cache, "fscache") != 0
		&& strcmp(cache, "mmap") != 0) {
			errx(1, "%s: unknown cache mode (expecting loose, fscache, or mmap)", cache);
		}
		snprintf(buf, sizeof(buf), "cache=%s", cache);
		append(&opts, buf, &optlen);
	}

	if (debugstr) {
		for (cp=strtok(debugstr, ","); cp; cp=strtok(NULL, ",")) {
			for (i=0; i<nelem(debug_flags); ++i) {
				if (strcmp(cp, debug_flags[i].mnemonic) == 0) {
					debug |= debug_flags[i].mask;
					break;
				}
			}
			if (i >= nelem(debug_flags)) {
				errx(1, "%s: unrecognised debug channel", cp);
			}
		}
		snprintf(buf, sizeof(buf), "debug=0x%04x", debug);
		append(&opts, buf, &optlen);
	}

	if (msize) {
		unsigned long nmsize;
		char *end = NULL;
		nmsize = strtoul(msize, &end, 10);
		if (*end || nmsize == 0 || nmsize > INT_MAX) {
			errx(1, "%s: msize must be a positive integer", msize);
		}
		if (pw->pw_uid != 0 && nmsize > Maxmsize) {
			nmsize = Maxmsize;
		}
		snprintf(buf, sizeof(buf), "msize=%lu", nmsize);
		append(&opts, buf, &optlen);
	}

	snprintf(buf, sizeof(buf), "name=%s", pw->pw_name);
	append(&opts, buf, &optlen);

	if (getenv("USER")) {
		snprintf(buf, sizeof(buf), "uname=%s", getenv("USER"));
	} else {
		snprintf(buf, sizeof(buf), "uname=%s", pw->pw_name);
	}
	if (strchr(buf, ',')) {
		errx(1, "%s: username can't contain commas", buf+6);
	}
	append(&opts, buf, &optlen);

	if (axess == -1) {
		append(&opts, "access=any", &optlen);
	} else if (axess) {
		snprintf(buf, sizeof(buf), "access=%d", axess);
		append(&opts, buf, &optlen);
	}
	if (!dotu) {
		append(&opts, "noextend", &optlen);
	}
	if (!dev) {
		append(&opts, "nodev", &optlen);
	}
	if (pw->pw_uid != 0) {
		append(&opts, "nosuid", &optlen);
	}
	if (uidgid) {
		snprintf(buf, sizeof(buf), "uid=%d,gid=%d", getuid(), getgid());
		append(&opts, buf, &optlen); /* < 2.6.24 */
		snprintf(buf, sizeof(buf), "dfltuid=%d,dfltgid=%d", getuid(), getgid());
		append(&opts, buf, &optlen); /* >= 2.6.24 */
	}
	if (port) {
		snprintf(buf, sizeof(buf), "port=%d", port);
		append(&opts, buf, &optlen);
	}

	if (strcmp(proto, "tcp") == 0) {
		struct addrinfo *ai;
		struct addrinfo aihints;
		int r;
		memset(&aihints, 0, sizeof(aihints));
		aihints.ai_family = AF_INET;
		aihints.ai_socktype = SOCK_STREAM;
		if ((r=getaddrinfo(addr, NULL, &aihints, &ai))) {
			errx(1, "getaddrinfo: %s", gai_strerror(r));
		}
		if ((r=getnameinfo(ai->ai_addr, ai->ai_addrlen, buf,
						sizeof(buf), NULL, 0, NI_NUMERICHOST))) {
			errx(1, "getnameinfo: %s", gai_strerror(r));
		}
	} else { /* unix socket, virtio device or fd transport */
		snprintf(buf, sizeof(buf), "%s", addr);
	}

	if(dryrun) {
		fprintf(stderr, "mount -t 9p -o %s %s %s\n", opts, buf, mountpt);
	} else if (mount(buf, mountpt, "9p", 0, (void*)opts)) {
		err(1, "mount");
	}

	return 0;
}
